packages/runtime/container-runtime/src/gc/gcTelemetryTracker.ts

```diff
@@ -0,0 +1,353 @@
+ */
+
+import { ITelemetryLogger } from "@fluidframework/common-definitions";
+import { IGarbageCollectionData } from "@fluidframework/runtime-definitions";
+import { packagePathToTelemetryProperty } from "@fluidframework/runtime-utils";
+import {
+	generateStack,
+	MonitoringContext,
+	TelemetryDataTag,
+} from "@fluidframework/telemetry-utils";
+import { ICreateContainerMetadata } from "../summary";
+import {
+	disableSweepLogKey,
+	GCNodeType,
+	UnreferencedState,
+	IGarbageCollectorConfigs,
+} from "./gcDefinitions";
+import { sendGCUnexpectedUsageEvent } from "./gcHelpers";
+import { UnreferencedStateTracker } from "./gcUnreferencedStateTracker";
+
+type NodeUsageType = "Changed" | "Loaded" | "Revived";
+
+/** The event that is logged when unreferenced node is used after a certain time. */
+interface IUnreferencedEventProps
+	extends ICreateContainerMetadata,
+		Omit<INodeUsageProps, "nodeId" | "currentReferenceTimestampMs" | "packagePath"> {
+	usageType: NodeUsageType;
+	state: UnreferencedState;
+	id: string;
+	type: GCNodeType;
+	unrefTime: number;
+	age: number;
+	timeout?: number;
+}
+
+/** Properties passed to nodeUsed function when a node is used. */
+interface INodeUsageProps {
+	nodeId: string;
+	usageType: NodeUsageType;
+	currentReferenceTimestampMs: number | undefined;
+	packagePath: readonly string[] | undefined;
+	completedGCRuns: number;
+	isTombstoned: boolean;
+	lastSummaryTime?: number;
+	fromId?: string;
+	viaHandle?: boolean;
+}
+
+/**
+ * Encapsulates the logic that tracks the various telemetry logged by the Garbage Collector. There are 4 types of
+ * telemetry logged:
+ * 1. inactiveObject telemetry - When an inactive node is used - A node that has been unreferenced for inactiveTimeoutMs.
+ * 2. sweepReadyObject telemetry - When a sweep ready node is used - A node that has been unreferenced for sweepTimeoutMs.
+ * 3. Tombstone telemetry - When a tombstoned node is used - A node that that has been marked as tombstone.
+ * 4. Sweep / deleted telemetry - When a node is detected as sweep ready in the sweep phase.
+ * 5. Unknown outbound reference telemetry - When a node is referenced but GC is not explicitly notified of it.
+ */
+export class GCTelemetryTracker {
+	// Keeps track of unreferenced events that are logged for a node. This is used to limit the log generation to one
+	// per event per node.
+	private readonly loggedUnreferencedEvents: Set<string> = new Set();
+	// Queue for unreferenced events that should be logged the next time GC runs.
+	private pendingEventsQueue: IUnreferencedEventProps[] = [];
+
+	constructor(
+		private readonly mc: MonitoringContext,
+		private readonly configs: Pick<
+			IGarbageCollectorConfigs,
+			"inactiveTimeoutMs" | "sweepTimeoutMs"
+		>,
+		private readonly isSummarizerClient: boolean,
+		private readonly gcTombstoneEnforcementAllowed: boolean,
+		private readonly createContainerMetadata: ICreateContainerMetadata,
+		private readonly getNodeType: (nodeId: string) => GCNodeType,
```

Comment:
Why does this need to support multiple implementations of getNodeType?

Reply:
I am not sure what you mean by that. What are the multiple implementations?

Reply:
Why is the function passed in instead of just implementing the functionality or using a shared helper.

Reply:
Ohh right I see, even GarbageCollector doesn't know how this is implemented, it comes from the Runtime like `(nodeId: string) => this.runtime.getNodeType(nodeId),`